Sorter performance

TEdwardJ opened the pull request.

TEdwardJ committed:
Sorter performance:
 - some extra method calls are eliminated;

TEdwardJ committed:
Sorter optimization:
 - currentMatching is evaluated by method now, the tests for method are added;
 - enclosure of if statements in the addOffersSortingByVinyl() method is decreased;

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/util/ParserConfiguration.java:
```diff
@@ -0,0 +1,19 @@
+    String highResolutionImageSelector;
+    String inStockMarkerSelector;
+    String inStockMarker;
+}
```
Please add empty rows after class declaration and before closing

AlexKievDeveloper commented on src/main/java/com/vinylteam/vinyl/util/impl/DetailedVinylParserImpl.java:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        if (imageLink != null && !Objects.equals(imageLink, "")){
+            if (!imageLink.contains("no-cover")){
```
Ctrl+Alt+L

AlexKievDeveloper commented on src/test/java/com/vinylteam/vinyl/util/ParserConfigurationTest.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Test
+    @DisplayName("Checks whether builder built configuration correctly")
+    void testParserConfigurationBuilder(){
```
Ctrl+Alt+L

Greengrer commented on src/main/java/com/vinylteam/vinyl/util/PriceUtils.java:
```diff
@@ -24,8 +24,8 @@ public static double getPriceFromString(String fullPriceDetails) {
     public static Optional<Currency> getCurrencyFromString(String fullPriceDetails) {
-        if (fullPriceDetails.length() > 0) {
-            String priceDetails = fullPriceDetails;
+        String priceDetails = fullPriceDetails.replaceAll(",", ".").replaceAll("[0123456789. ]", "");
```
The reason I avoid regex as much as I can is because using regular expressions makes code less readable and harder to understand for hypothetical junior that has just joined the project.

It is just a note, I don't request changes here.

Greengrer commented on src/main/java/com/vinylteam/vinyl/util/RawOffersSorter.java:
```diff
@@ -46,50 +46,57 @@
+        int percentMatching = 75;
+        String uniqueVinylRelease = uniqueVinyl.getRelease();
+        char lastCharInRawRelease = uniqueVinylRelease.charAt(uniqueVinylRelease.length() - 1);
+        if (lastCharInRawRelease > '0' && lastCharInRawRelease < '9') {
+            percentMatching = 90;
+        }
```
"What does it mean? character > character, chatacter < character. Why percent matching jumps to 90?" 
Junior's perspective.

TEdwardJ replied:
Sorry, it is just the way the author conceived it. You`d better ask Taras about it, but as Taras explained, it is some special condition for release with numbers in its name, e.g. "Oxygène 3"

Greengrer commented on src/main/java/com/vinylteam/vinyl/util/RawOffersSorter.java:
```diff
@@ -46,50 +46,57 @@
+            percentMatching = 90;
+        }
+        String[] preparedFullNameForMatching = Arrays
+                .stream(uniqueVinyl.getFullName().split("[- ()!@$%^&*_={}:;\"']"))
```
Regexes that are hard to understand.
Not requesting change just noting.

Greengrer commented on src/main/java/com/vinylteam/vinyl/util/RawOffersSorter.java:
```diff
@@ -46,50 +46,57 @@
+        while (rawOfferIterator.hasNext()) {
+            RawOffer rawOffer = rawOfferIterator.next();
+            String rawOfferFullName = rawOffer.getArtist() + " - " + rawOffer.getRelease();
+            int currentMatching = getMatchNumber(preparedFullNameForMatching, rawOfferFullName);
```
Why not make getMatchNumber return float?

Greengrer replied:
Could you change names:
percentMatching = minMatchingRate;
currentMatching = currentMatchingRate.
Make getMatchNumber return percentage of matching text instead of the number of characters that match - getMatchingRate?
And put currentMatchingRate inside if(),
Then it could be like that:

```
int(or float) minMatchingRate = 75;
...
if (Objects.equals(preparedVinylRelease,...) && Object.equals(...)) {
float currentMatchingRate = getMatchingRate(preparedFullNameForMatching, rawOfferFullName);
if (currentMatchingRate > minMatchingRate) {
//adding offer, removing RawOffer.
}
}
```

TEdwardJ replied:
agreed, getMatchingRate() method is created, moreover we can group these two if  statements into one...
But:
 - currentMatching  - it is not about rate, it contains just counting result, so I would name it currentMatchingNumber;
 - I would leave the logic for rate evaluation and for counting of matches divided, just in order to follow single responsibility principle and to make the possible changes in formula and in counting more easier, so there is one more method, special for counting: getMatchingNumber; moreover it is easier to test the logic;

Greengrer commented on src/main/java/com/vinylteam/vinyl/util/RawOffersSorter.java:
```diff
@@ -46,50 +46,57 @@
+                }
+            }
+        }
+        uniqueVinyl.setHasOffers(!offers.isEmpty());
```
You might pass not empty list of offers in this method.
That means that when you don't find any offers for passed unique vinyl and you set uniqueVinyl.setHasOffers(!offers.isEmpty) hasOffers becomes true.

Move uniqueVinyl.setHasOffers() inside if in 75-88.
That means you don't have tests that test case when
[uniqueVinyl1(true, some rawOffers match), uniqueVinyl2(true, noRawOffersMatch)] -> sorter -> [uniqueVinyl1(true, some offers), uniqueVinyl2(false, no offers)], [offer1(for uniqueVinyl1)].

TEdwardJ replied:
"That means that when you don't find any offers for passed unique vinyl and you set uniqueVinyl.setHasOffers(!offers.isEmpty) hasOffers becomes true" - incorrect suggestion, it becomes false
...
seems I got it, thanks! to be reworked!

Greengrer commented on src/main/java/com/vinylteam/vinyl/util/RawOffersSorter.java:
```diff
@@ -46,50 +46,57 @@
+        uniqueVinyl.setHasOffers(!offers.isEmpty());
+    }
+
+    int getMatchNumber(String[] preparedFullNameForMatching, String rawOfferFullName) {
```
float getMatchingRate()?

TEdwardJ replied:
good idea, to create method that would be responsible for rate evaluation. This one is responsible only for matching counting

Greengrer commented on src/main/java/com/vinylteam/vinyl/util/RawOffersSorter.java:
```diff
@@ -46,50 +46,57 @@
+
+    int getMatchNumber(String[] preparedFullNameForMatching, String rawOfferFullName) {
+        return (int) Arrays.stream(preparedFullNameForMatching)
+                .map(preparedItem -> preparedItem.toLowerCase().trim())
```
You don't need trim after trimming each element on line 63.

TEdwardJ replied:
there is just a filter in line 63, it does not change element that passed the filter

Greengrer commented on src/main/java/com/vinylteam/vinyl/util/impl/DeejayDeParser.java:
```diff
@@ -0,0 +1,158 @@
+        return offerLinks;
+    }
+
+    Set<String> getAllPages(String baseLink) {
+        Set<String> linkSet = new HashSet<>();
+        return getAllPages(linkSet, baseLink);
+    }
+
+    Set<String> getAllPages(Set<String> linkSet, String baseLink) {
```
Why do you make two?

TEdwardJ replied:
the second of these overloaded methods are called recursively, so one general collection was required in order to accumulate all the data

Greengrer commented on src/test/java/com/vinylteam/vinyl/util/RawOffersSorterTest.java:
```diff
@@ -242,6 +247,24 @@ void getParametersForComparisonWhenStringContainsManyWordsWithoutArticleTest() {
+    @MethodSource("getTestData")
+    @DisplayName("Checks whether match counter works correctly")
+    void getMatchNumberTest(int expectedNumber, String[] searchWords, String rawOfferFullName) {
+        int number = sorter.getMatchNumber(searchWords, rawOfferFullName);
```
getMatchingRate?

Greengrer replied:
Shouldn't expectedNumbers be set within the test?
number - actualNumber, but I'm just cherrypicking.

TEdwardJ replied:
getMatchingRate - good name, agreed, but the method does not return rate, the rate is evaluated later.It just returns number of matching

Greengrer commented on src/test/java/com/vinylteam/vinyl/util/impl/DeejayDeParserTest.java:
```diff
@@ -0,0 +1,82 @@
+        RawOffer rawOffer = parser.getRawOfferFromElement(vinylElement, parser.getOnePageDetailedParser());
+        assertNotNull(rawOffer);
+        assertEquals("Dixon Ep", rawOffer.getRelease());
+        //assertTrue(rawOffer.getOfferLink().isEmpty());
```
commented  code.

Greengrer reviewed (changes requested):
In the sorter there is a uniqueVinyl.setHasOffers(!offers.isEmpty()) after sorting all raw offers by that unique vinyl. If list offers passed to the method isn't empty initially, it might assign hasOffers==true to the unique vinyl that didn't have any matches within rawOffers.

TEdwardJ committed:
Sorter optimization Pull Request, remarks are taken into account:
 - variable names changed;
 - some new methods were extracted from general flow of the code with the appropriate tests;

TEdwardJ committed:
Merge branch 'develop' of https://github.com/littlepony99/wax-deals into sorter_performance

CodeClub2021-yul commented on src/main/java/com/vinylteam/vinyl/util/RawOffersSorter.java:
```diff
@@ -46,51 +46,68 @@
-        } else {
+        if (rawOffers == null || uniqueVinyl == null || offers == null) {
             RuntimeException e = new NullPointerException();
             log.error("At least one of passed arguments is null {'rawOffers':{}, 'uniqueVinyl':{}, 'offers':{}}",
```
I think it is not good idea to throw NullPointerException

Yuliia-Sh commented on src/main/java/com/vinylteam/vinyl/util/RawOffersSorter.java:
```diff
@@ -46,51 +46,68 @@
             log.error("At least one of passed arguments is null {'rawOffers':{}, 'uniqueVinyl':{}, 'offers':{}}",
                     rawOffers, uniqueVinyl, offers, e);
             throw e;
         }
```
I think it is not good idea to throw NullPointerException

TEdwardJ replied:
Actually this check is redundant at all, no chance to pass null in any of parameters - collections are initialized before in program flow, uniqueVinyl is taken from iterator, and in case iterator does not have any more vinyls, the special logic adds new one - to be reworked

TEdwardJ committed:
Sorter Performance review:
 - formatting is applied where it was missed;
 - RawOfferSorter.addOffersSortingByVinyl() has redundant check for nulls eliminated;
